0027-safely-recommend-actions.adoc

```diff
@@ -0,0 +1,173 @@
+----
+
+Please note, that the `TTI` need not be the same for the redeem transaction as for the fund transaction since different fees can be used for each transaction as required.
+
+== Functionality
+
+The system is expected to provide the following functionality:
+
+.Requirements
+1. Given two blockchains, create a pair of useful expiries.
+2. Given a pair of expiries and their respective blockchains, determine if the expiries are useful.
+3. Given the current swap state, determine what the user should do.
+
+A pair of expiries is 'useful' if, at this moment, both parties are able to execute all steps to complete the swap.
+
+== Solution
+
+Here we describe solutions to each of the listed requirements above.
+
+1. To create a pair of useful expiries we calculate the expected time required for completion of the swap. That is the sum of the transition periods for each transition. We then add a Time To Live (`TTL`) value, this ensures the calculated expiries remain useful for `TTL`.
+
+2. A pair of expiries is useful if there is time to take each action required to complete the swap i.e., the expiry pair is useful if there is time to transition to the first action AND there is time for each subsequent action.
+
+3. Semantically the next possible action is either the next happy path action or cancel. Cancel is either refund or abort.
+
+.Semantic actions
+* *next*: The next happy path action (deterministic from the state diagram).
+* *refund*: Send a refund transaction to the blockchain (cancel after fund).
+* *wait_for_refund*: Refund is not yet available but should be taken when available.
+* *abort*: Take no further action (cancel before fund).
+
+The recommended next action can be fully determined from the following algorithm:
+
+----
+1. IF refund is available RETURN refund
+2. IF there is time to transition to next AND there is time for the subsequent action RETURN next
+3. IF have funded RETURN wait_for_refund
+4. RETURN abort
+----
+
+Note that step 2 is recursive, the base case being the final stage of the annotated state diagram - done (see below).
+
+=== Explanation
+
+Various events must occur in order to transition from one state to the next. We can identify these events and annotate the state diagram accordingly. If we are able to calculate the time of each of these events we are able to fully determine the period of each transition.
+
+We define the Time To Final to be a part of the transition to a state e.g., the alpha funded state is not reached until the alpha fund transaction has become final.
+
+There exists a transition diagram for each actor. Each actor progresses along their own state diagram but must be able to predict the state progression of the counterparty in order to determine whether progression is possible for the counterparty.
+
+.Annotated state diagram for Alice:
+----
+(Create) TTA_A TTI_AF TTF_A (alpha funded) TTA_B TTI_BF TTF_B (beta funded) TTA_A TTI_BR TTC_B (beta redeemed) (done)
+----
```

Comment:
Sorry, I have no idea what that means :) If this is some kind of standard, can you link the source to the model please?

Reply:
I don't really know how to make this clearer. The acronyms represent functions that return a time value, these state diagrams include the time values required to transition from one state to the next. If there is some standard for writing this I'm happy to learn it.